### ISSUE TITLE
not able to go to definition of non exported functions belonging to same package

### ISSUE THREAD

**Issue by vigneshkm** [fatih/vim-go#2188]
### What did you do? (required. The issue will be **closed** when not provided.)
I am trying to go to definition of functions with Ctrl-] key. I am able to find definition of Exported functions (beginning with Capital letter) but not able to go to definition of non-exported functions belonging to same package.

eg:
1) Able to goto definition of context() in `sdk.Context()` from other packages.
2) Not able to go to definition of newIdentity() in `sdk.newIdentity()` from within the package containing newIdentity.

### What did you expect to happen?
Go to definition of non exported functions in a package.

### What happened instead?
Seeing below error:
`vim-go: [searching declaration] FAIL`

### Configuration (**MUST** fill this out):

* vim-go version:
master

* `vimrc` you used to reproduce (use a *minimal* vimrc with other plugins disabled; do not link to a 2,000 line vimrc):
" A sensible vimrc for Go development
"
" Please note that the following settings are some default that I used
" for years. However it might be not the case for you (and your
" environment). I highly encourage to change/adapt the vimrc to your own
" needs. Think of a vimrc as a garden that needs to be maintained and fostered
" throughout years. Keep it clean and useful - Fatih Arslan

call plug#begin()
Plug 'fatih/vim-go'
Plug 'fatih/molokai'
Plug 'AndrewRadev/splitjoin.vim'
Plug 'SirVer/ultisnips'
Plug 'ctrlpvim/ctrlp.vim'
call plug#end()

""""""""""""""""""""""
"      Settings      "
""""""""""""""""""""""
set nocompatible                " Enables us Vim specific features
filetype off                    " Reset filetype detection first ...
filetype plugin indent on       " ... and enable filetype detection
set ttyfast                     " Indicate fast terminal conn for faster redraw
set ttymouse=xterm2             " Indicate terminal type for mouse codes
set ttyscroll=3                 " Speedup scrolling
set laststatus=2                " Show status line always
set encoding=utf-8              " Set default encoding to UTF-8
set autoread                    " Automatically read changed files
set autoindent                  " Enabile Autoindent
set backspace=indent,eol,start  " Makes backspace key more powerful.
set incsearch                   " Shows the match while typing
set hlsearch                    " Highlight found searches
set noerrorbells                " No beeps
set number                      " Show line numbers
set showcmd                     " Show me what I'm typing
set noswapfile                  " Don't use swapfile
set nobackup                    " Don't create annoying backup files
set splitright                  " Vertical windows should be split to right
set splitbelow                  " Horizontal windows should split to bottom
set autowrite                   " Automatically save before :next, :make etc.
set hidden                      " Buffer should still exist if window is closed
set fileformats=unix,dos,mac    " Prefer Unix over Windows over OS 9 formats
set noshowmatch                 " Do not show matching brackets by flickering
set noshowmode                  " We show the mode with airline or lightline
set ignorecase                  " Search case insensitive...
set smartcase                   " ... but not it begins with upper case
set completeopt=menu,menuone    " Show popup menu, even if there is one entry
set pumheight=10                " Completion window max size
set nocursorcolumn              " Do not highlight column (speeds up highlighting)
set nocursorline                " Do not highlight cursor (speeds up highlighting)
set lazyredraw                  " Wait to redraw

" Enable to copy to clipboard for operations like yank, delete, change and put
" http://stackoverflow.com/questions/20186975/vim-mac-how-to-copy-to-clipboard-without-pbcopy
if has('unnamedplus')
  set clipboard^=unnamed
  set clipboard^=unnamedplus
endif

" This enables us to undo files even if you exit Vim.
if has('persistent_undo')
  set undofile
  set undodir=~/.config/vim/tmp/undo//
endif

" Colorscheme
syntax enable
set t_Co=256
let g:rehash256 = 1
let g:molokai_original = 1
colorscheme molokai

""""""""""""""""""""""
"      Mappings      "
""""""""""""""""""""""

" Set leader shortcut to a comma ','. By default it's the backslash
let mapleader = ","

" Jump to next error with Ctrl-n and previous error with Ctrl-m. Close the
" quickfix window with <leader>a
map <C-n> :cnext<CR>
map <C-m> :cprevious<CR>
nnoremap <leader>a :cclose<CR>

" Visual linewise up and down by default (and use gj gk to go quicker)
noremap <Up> gk
noremap <Down> gj
noremap j gj
noremap k gk

" Search mappings: These will make it so that going to the next one in a
" search will center on the line it's found in.
nnoremap n nzzzv
nnoremap N Nzzzv

" Act like D and C
nnoremap Y y$

" Enter automatically into the files directory
autocmd BufEnter * silent! lcd %:p:h

"""""""""""""""""""""
"      Plugins      "
"""""""""""""""""""""

" vim-go
let g:go_fmt_command = "goimports"
let g:go_autodetect_gopath = 1
let g:go_list_type = "quickfix"

let g:go_highlight_types = 1
let g:go_highlight_fields = 1
let g:go_highlight_functions = 1
let g:go_highlight_function_calls = 1
let g:go_highlight_extra_types = 1
let g:go_highlight_generate_tags = 1

" Open :GoDeclsDir with ctrl-g
nmap <C-g> :GoDeclsDir<cr>
imap <C-g> <esc>:<C-u>GoDeclsDir<cr>

augroup go
  autocmd!

  " Show by default 4 spaces for a tab
  autocmd BufNewFile,BufRead *.go setlocal noexpandtab tabstop=4 shiftwidth=4

  " :GoBuild and :GoTestCompile
  autocmd FileType go nmap <leader>b :<C-u>call <SID>build_go_files()<CR>

  " :GoTest
  autocmd FileType go nmap <leader>t  <Plug>(go-test)

  " :GoRun
  autocmd FileType go nmap <leader>r  <Plug>(go-run)

  " :GoDoc
  autocmd FileType go nmap <Leader>d <Plug>(go-doc)

  " :GoCoverageToggle
  autocmd FileType go nmap <Leader>c <Plug>(go-coverage-toggle)

  " :GoInfo
  autocmd FileType go nmap <Leader>i <Plug>(go-info)

  " :GoMetaLinter
  autocmd FileType go nmap <Leader>l <Plug>(go-metalinter)

  " :GoDef but opens in a vertical split
  autocmd FileType go nmap <Leader>v <Plug>(go-def-vertical)
  " :GoDef but opens in a horizontal split
  autocmd FileType go nmap <Leader>s <Plug>(go-def-split)

  " :GoAlternate  commands :A, :AV, :AS and :AT
  autocmd Filetype go command! -bang A call go#alternate#Switch(<bang>0, 'edit')
  autocmd Filetype go command! -bang AV call go#alternate#Switch(<bang>0, 'vsplit')
  autocmd Filetype go command! -bang AS call go#alternate#Switch(<bang>0, 'split')
  autocmd Filetype go command! -bang AT call go#alternate#Switch(<bang>0, 'tabe')
augroup END

" build_go_files is a custom function that builds or compiles the test file.
" It calls :GoBuild if its a Go file, or :GoTestCompile if it's a test file
function! s:build_go_files()
  let l:file = expand('%')
  if l:file =~# '^\f\+_test\.go$'
    call go#test#Test(0, 1)
  elseif l:file =~# '^\f\+\.go$'
    call go#cmd#Build(0)
  endif
endfunction

* Vim version (first three lines from `:version`):
VIM - Vi IMproved 8.1 (2018 May 18, compiled Mar 19 2019 10:15:10)

* Go version (`go version`):
go version go1.9.2 linux/amd64

* Go environment (`go env`):
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/vignesh/go-book/"
GORACE=""
GOROOT="/usr/local/go"
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build391835598=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"

**Comment by bhcleek** [comment]
I've tried this using all three supported values for `g:go_def_mode`, and it works in each of them for me. The only thing I can think of is that your version of Go is pretty old, so you may also need to use some older versions of tools like `godef`.

The best way to troubleshoot this is to set `g:go_debug=['shell-commands']` and then try jumping to an unexported identifier. Then, you can run `:messages` to see the command that vim-go used to try the jump. Take that command, run it on the command line manually and see what the output of the tools is.

**Comment by vigneshkm** [comment]
@bhcleek , I am using go1.12.1. Also I could not get what you meant with `'shell-commands'` in `g:go_debug=['shell-commands']`. Following is the recording of my screen where you could see the exact issue that I am facing.
[https://www.youtube.com/watch?v=V3zSZvdQsJA&feature=youtu.be](https://www.youtube.com/watch?v=V3zSZvdQsJA&feature=youtu.be)

**Comment by bhcleek** [comment]
Have you upgraded Go since you created this issue (your original report clearly says you were using Go 1.9)? If so, have you tried running `:GoUpdateBinaries ` since upgrading?

**Comment by vigneshkm** [comment]
When I reported the issue, I was using go1.9. Now I am using go1.12.1. After your suggestion I did `:GoUpdateBinaries` but still finding the issue.

**Comment by bhcleek** [comment]
The video you linked, https://www.youtube.com/watch?v=V3zSZvdQsJA&feature=youtu.be, is not publicly available.

I've tried duplicating what you've described without success.

> Also I could not get what you meant with 'shell-commands' in g:go_debug=['shell-commands']

Literally, run `let g:go_debug=['shell-commands']` in Vim before trying to jump to the definition.

**Comment by vigneshkm** [comment]
I have now changed the video permission to public and you could watch it now..
[https://www.youtube.com/watch?v=V3zSZvdQsJA&feature=youtu.be](https://www.youtube.com/watch?v=V3zSZvdQsJA&feature=youtu.be)
I also tried to run let `g:go_debug=['shell-commands']` and it to  gave same error. You could see the error at [https://youtu.be/O3SDz12o7Tg](https://youtu.be/O3SDz12o7Tg)

**Comment by bhcleek** [comment]
Thank you for making the video public. Unfortunately, it didn't offer any additional insight over what you've already described. The best way forward is to use the instructions I provided previously to get more information about what's happening:

> set g:go_debug=['shell-commands'] and then try jumping to an unexported identifier. Then, you can run :messages to see the command that vim-go used to try the jump. Take that command, run it on the command line manually and see what the output of the tools is.

**Comment by bhcleek** [comment]
Your second video may actually be helpful. I noticed that it looks like the package you're working in is not in your $GOPATH, but it's not clear whether it's a module either. Are you working in module mode or GOPATH mode?

**Comment by vigneshkm** [comment]
At $GOPATH, I was having a soft-link to the package but not the package itself. Now I copied the package into $GOPATH,  and it is working.. I am able to go to definition of unexported fields. 
Thanks a lot for the suggestion. 
Want to know one more thing. I had cloned the package outside $GOPATH and while compiling, I used the softlink. With Softlink, I could compile the package properly. So, is it not correct to have softlinks??
